### PR TITLE
refactor(core): css-custom-property feature extraction

### PR DESCRIPTION
**Pull request by idoros**
## fix
- [x] questioned in #1954  currently `@st-global-custom-property` overrides the `@property` definition regardless of the definitions order [see behavior](https://wixplosives.github.io/stylable-playground/#state=%28%27files%21%282%21%27085865la10la7%27%29%7Eselected%212%27%29*-property%2520--30%25403cu4om*1762%27%252Findex.4.css34-global-4st5%25404*6%255Cn74%253B8fir1%2501876543210*_) - This PR resolves and tests the definition behavior: 
  - deprecated with a deprecation info diagnostic
  - redeclare diagnostic when conflicting with `@property` and **mark property as global**
  - redeclare diagnostic when conflicting with `@st-import` and override with global symbol
  - define global property symbol in any other case
- [x] imported properties are not collected during analyze, and `prop/var()` usages creates an incorrect symbol - this PR fixes the issue by adding an `alias` ref to the imported symbol (like `class` and `type` symbols), and imported custom properties are identified by the `--` prefix and added to as an aliased property
- [x] empty `var()` break the process with: [`cannot read properties of undefined (reading 'type')`](https://wixplosives.github.io/stylable-playground/#state=%28%27files%21%28*%21%27.root%2520%28%255Cn%2520%2520%2520%2520color%253A%2520var%257B%257D%253B%255Cn%29%27%29%7Eselected%21*%27%29*%27%252Findex.st.css%2501*_) - this PR keeps `var()` untransformed, report `missing property name` diagnostic and doesn't crush
- noticed on the way
  - [x] fix(core): wrong transform for empty `@keyframes {}` -> `@keyframesNS__ {}` [see playground](https://wixplosives.github.io/stylable-playground/#state=%28%27files%21%28*%21%27%2540keyframes%2520%28%29%27%29%7Eselected%21*%27%29*%27%252Fbutton.st.css%2501*_)
## refactor
- processor
  - [x] simplify the property collection process
  - [x] `analyzeAtRule`
    - move definition/validation of `@property` to feature
    - move definition of `@st-global-custom-property` to feature
  - [x] `analyzeDeclaration` (new hook)
    - move "declare-unknown" definition for `--X: value;` to feature
    - move "declare-unknown" definition for `prop: var(--X)` to feature
- transformer
  - [x] `transformResolve` - move `createCSSVarsMapping` to feature
  - [x] `transformAtRule` - move `@property` namespace/removal to feature
  - [x] `transformDeclaration` - move `--X: ;` namespace to feature
  - [x] `transformDeclarationValue` (new hook) - move `: var();` namespace to feature _(value transformation is part of `evelDeclarationValue`)_
  - [x] `transformJSExports` - move `exportCSSVars` to feature
- small
  - [x] `CSSVarSymbol` to feature
  - [x] move `validate-at-property` to `helpers/css-custom-property`
  - [x] move code to new `helpers/global`
  - [x] move `stylable-utils/isCSSVarProp` to `helpers/css-custom-property`

## deprecate
- [x] `meta.cssVars`
- [x] `paramMapping` export

## test
- [x] move tests to `css-property.spec`
- [x] add missing tests
- fix(core-test-kit): inline expectation `@analayze` & `@transform` bug with multiple lines

### CODE REVIEW

File: packages/core/src/stylable-transformer.ts [hunk 1]
```diff
@@ -278,64 +299,16 @@ export class StylableTransformer {
...
-                cssVarsMapping[localVarName] = cssVar.global
-                    ? localVarName
-                    : generateScopedCSSVar(meta.namespace, localVarName.slice(2));
-            }
-        }
-
-        return cssVarsMapping;
-    }
+    /** @deprecated */
     public getScopedCSSVar(
```
Comment (barak007) [inline packages/core/src/stylable-transformer.ts:303]: deprecate

File: packages/core/test/features/css-custom-property.spec.ts [hunk 2]
```diff
@@ -0,0 +1,869 @@
...
+                /* @decl(basic) prop: var(--entry-colorA) */
+                prop: var(--colorA);
+
+                /* @decl(concat) prop: var(--entry-colorA) var(--entry-colorB) */
+                prop: var(--colorA) var(--colorB);
+
+                /* @decl(within value) prop: 2px var(--entry-colorB) solid */
+                prop: 2px var(--colorB) solid;
+
+                /* @decl(as default) prop: var(--entry-colorC, black) */
```
Comment (barak007) [inline packages/core/test/features/css-custom-property.spec.ts:61]: with default

File: packages/core/test/features/css-custom-property.spec.ts [hunk 3]
```diff
@@ -0,0 +1,869 @@
...
+            }
+        `);
+
+        const { meta } = sheets['/entry.st.css'];
+
+        shouldReportNoDiagnostics(meta);
+    });
+    it(`should collect global css props`, () => {
+        const { sheets } = testStylableCore(`
+            @property st-global(--propX);
```
Comment (barak007) [inline packages/core/test/features/css-custom-property.spec.ts:190]:
```suggestion
            /* @transform-remove(only type) */
            @property st-global(--propX);
```

**Review by barak007 — CHANGES_REQUESTED** [review]

File: packages/core/src/features/css-custom-property.ts [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+                    symbol: cssVar,
+                    meta,
+                });
+            }
+        }
+
+        return customPropsMapping;
+    },
+    transformAtRuleNode({ atRule, resolved }) {
+        if (atRule.nodes?.length) {
```
[the remark this answers is not in the document]
Comment (idoros) [reply packages/core/src/features/css-custom-property.ts:159]: declaration prop and value don't are general and probably don't need it

File: packages/core/src/stylable-transformer.ts (hunk 1 above)
Comment (idoros) [reply packages/core/src/stylable-transformer.ts:303]: ???

File: packages/cli/src/code-mods/st-global-custom-property-to-at-property.ts [hunk 5]
```diff
@@ -1,10 +1,6 @@
-import {
-    CSSVarSymbol,
-    Diagnostics,
-    isCSSVarProp,
-    paramMapping,
-    processorWarnings,
-} from '@stylable/core';
+import { CSSVarSymbol, Diagnostics, isCSSVarProp } from '@stylable/core';
+import { GLOBAL_FUNC } from '@stylable/core/dist/helpers/global';
```
[the remark this answers is not in the document]
Comment (idoros) [reply packages/cli/src/code-mods/st-global-custom-property-to-at-property.ts:2]: removed this specifically, although this is the way we pass internals between packages until we use package.json exports

File: packages/core/src/features/css-custom-property.ts [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+    transformDeclarationValue({ node, resolved }) {
+        const { value } = node;
+        const varWithPrefix = node.nodes[0]?.value || ``;
+        if (isCSSVarProp(varWithPrefix)) {
+            if (resolved && resolved[varWithPrefix]) {
+                node.nodes[0].value = resolved[varWithPrefix];
+            }
+        }
+        // handle default values
+        if (node.nodes.length > 2) {
```
[the remark this answers is not in the document]
Comment (idoros) [reply packages/core/src/features/css-custom-property.ts:183]: this code has been like this for a while and is a combination from several PRs:
https://github.com/wix/stylable/commit/dc54b2183475f72b0d1e6cdd82a3b2554518a9df
https://github.com/wix/stylable/commit/cf8d5defa135303198700681702085d4ccefe4b3
https://github.com/wix/stylable/commit/0fe6a97cc65676988082e31c4c3d8ecea04d23cc
https://github.com/wix/stylable/commit/dc54b2183475f72b0d1e6cdd82a3b2554518a9df

Comments have no affect over this flow, it just "stringifyFunction" the entire function if there is more the the property name passed to the `var` function, which is a process that happens with or without this code. I left a ToDo comment to check if this is required as I couldn't figure it out.

**Review by barak007 — APPROVED** [review]

**Pull request merged**